<a href="https://colab.research.google.com/github/GameboyZero/404/blob/master/%D0%A2%D0%B5%D1%81%D1%82%D0%BE%D0%B2%D0%BE%D0%B5%20%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%20-%20%D0%BA%D0%BB%D0%B0%D1%81%D1%81%D0%B8%D1%84%D0%B8%D0%BA%D0%B0%D1%82%D0%BE%D1%80%20%D0%B1%D0%B0%D0%BD%D0%BA%D0%BD%D0%BE%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Описание задачи: В предложенном датасэте извлеченными признаками являются дисперсия (VAR), асимметрия(SKE), эксцесс(KUR) и Энтропия(ENT) - у меня в таблице эти признаки записаны как V1,V2,V3,V4. Набор данных состоит только из этих четырех признаков и меток классов. Набор данных был задуман для того, чтобы различать поддельные и подлинные банкноты.  Из 1372 экземпляров 610 кодируются как 1, обозначающий подлинные банкноты, а остальные-как 0

Задача: написать классификатор

In [0]:
from google.colab import files
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier



In [3]:
# Загружаем базу

files.upload()
!ls

Saving banknote.csv to banknote.csv
banknote.csv  sample_data


In [0]:
# Читаем базу

df = pd.read_csv("banknote.csv") 

In [43]:
# Выводим названия полей 

df.head()

V1      V2      V3       V4  Class
0  3.62160  8.6661 -2.8073 -0.44699      1
1  4.54590  8.1674 -2.4586 -1.46210      1
2  3.86600 -2.6383  1.9242  0.10645      1
3  3.45660  9.5228 -4.0112 -3.59440      1
4  0.32924 -4.4552  4.5718 -0.98880      1

In [44]:
df.isnull().sum()

V1       0
V2       0
V3       0
V4       0
Class    0
dtype: int64

In [45]:
df['Class'].unique()



array([1, 2])

In [46]:
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
V1       1372 non-null float64
V2       1372 non-null float64
V3       1372 non-null float64
V4       1372 non-null float64
Class    1372 non-null int64
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


In [47]:
df['Class'].value_counts()

1    762
2    610
Name: Class, dtype: int64

In [0]:
x = df.iloc[:,:-1]

In [49]:
x.head()

V1      V2      V3       V4
0  3.62160  8.6661 -2.8073 -0.44699
1  4.54590  8.1674 -2.4586 -1.46210
2  3.86600 -2.6383  1.9242  0.10645
3  3.45660  9.5228 -4.0112 -3.59440
4  0.32924 -4.4552  4.5718 -0.98880

In [0]:
y = df.iloc[:,-1]

In [51]:
y.head()



0    1
1    1
2    1
3    1
4    1
Name: Class, dtype: int64

20% примеров в наборе данных были использованы для тестирования, а остальные 80% были использованы для обучения модели. Теперь данные разделены в соотношении 80: 20 как набор для обучения и тестирования соответственно. Затем выбирается соответствующий алгоритм, обучаются и гиперпараметры настраиваются, чтобы дать лучшую модель. Затем модель будет проверена на данных испытаний, и если точность удовлетворительна, мы можем объявить модель точной, иначе мы перейдем к поиску другого подходящего классификатора.

In [0]:
# Разделяем датасэт на обучающий и тестовый набор данных
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [62]:
# Вывожу размер обучающей выборки

print(x_train.shape) 

(1097, 4)


In [63]:
# Вывожу размер проверочной выборки

print(x_test.shape) 

(275, 4)


In [64]:
# Из библиотеки sklearn использую кассификатор Случайный Лес

rf = RandomForestClassifier()
rf.fit(train_x,train_y)
pred_rf = rf.predict(test_x)
print(accuracy_score(test_y,pred_rf))

0.9927272727272727


Результат 0.9927272727272727 близок к 1,0 но всё же это не единица


In [40]:
# Из библиотеки sklearn использую кассификатор k-ближайших соседей

knn = KNeighborsClassifier(n_neighbors =3)
knn.fit(train_x, train_y)

Knn_pred = knn.predict(test_x)
accuracy_score(test_y, Knn_pred)

1.0

Метод KNN (k-ближайших соседей) показал более высокую точность, ровно 1,0